src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,403 @@
+                    compoundScores[subQueryNumber],
+                    collectedHits[subQueryNumber],
+                    sort.getSort()
+                )
+            );
+        }
+        return topFieldDocs;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return hitsThresholdChecker.scoreMode();
+    }
+
+    protected abstract class HybridTopDocSortLeafCollector implements LeafCollector {
+        protected HybridQueryScorer compoundQueryScorer;
+        private boolean collectedAllCompetitiveHits = false;
+        @Nullable
+        private FieldDoc after;
+        private final Sort sort;
+        private boolean initializePerSegment;
+
+        public HybridTopDocSortLeafCollector(Sort sort, @Nullable FieldDoc after) {
+            this.sort = sort;
+            this.after = after;
+            this.initializePerSegment = true;
+        }
+
+        @Override
+        public void setScorer(final Scorable scorer) throws IOException {
+            if (scorer instanceof HybridQueryScorer) {
+                log.debug("passed scorer is of type HybridQueryScorer, saving it for collecting documents and scores");
+                compoundQueryScorer = (HybridQueryScorer) scorer;
+            } else {
+                compoundQueryScorer = getHybridQueryScorer(scorer);
+                if (Objects.isNull(compoundQueryScorer)) {
+                    log.error(
+                        String.format(Locale.ROOT, "cannot find scorer of type HybridQueryScorer in a hierarchy of scorer %s", scorer)
+                    );
+                }
+            }
+        }
+
+        private HybridQueryScorer getHybridQueryScorer(final Scorable scorer) throws IOException {
+            if (scorer == null) {
+                return null;
+            }
+            if (scorer instanceof HybridQueryScorer) {
+                return (HybridQueryScorer) scorer;
+            }
+            for (Scorable.ChildScorable childScorable : scorer.getChildren()) {
+                HybridQueryScorer hybridQueryScorer = getHybridQueryScorer(childScorable.child);
+                if (Objects.nonNull(hybridQueryScorer)) {
+                    log.debug(
+                        String.format(
+                            Locale.ROOT,
+                            "found hybrid query scorer, it's child of scorer %s",
+                            childScorable.child.getClass().getSimpleName()
+                        )
+                    );
+                    return hybridQueryScorer;
+                }
+            }
+            return null;
+        }
+
+        /*
+        Increment total hit count and validate if threshold is reached.
+         */
+        protected void incrementTotalHitCount() throws IOException {
+            totalHits++;
+            hitsThresholdChecker.incrementHitCount();
+            if (scoreMode().isExhaustive() == false
```

Comment:
Can we remove notifying other comparators from this function and keep this function only to increasecount, Right now this functionName and responsibility does not match . 
Other reason is , in If Statement , there is 3 checks we are performing and that's going to happen every time when we Increment . If we can early terminate also that could be beneficial with that we wil saving some operations in if statement.

Reply:
No, we can't do that because this operation has to performed after incrementing total hits. 
Check this out https://github.com/apache/lucene/blob/057cbf3c8665354cd21dac20570a7e088dc8ce3a/lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java#L74